File: python/shpToGeojson/convertGeodetic.py
import json
from pathlib import Path

# angle units in degrees
def tky2wgs_approx(lon_tky, lat_tky, degree=2):
    L = lon_tky - 135
    B = lat_tky - 35
    if degree == 1:
        dL = -2.80641517e-3 - 4.98891541e-5*B - 8.33469263e-5*L
        dB =  3.20321021e-3 - 1.15452990e-4*B + 2.47032059e-5*L
    elif degree == 2:
        a00,a01,a02,a03,a04,a05,a10,a11,a12,a13,a14,a15 = -2.79648156e-03, -3.64571151e-05, -1.00958714e-06, -8.83091873e-05, -8.33002662e-07, 3.54561248e-07, 3.19774649e-03, -1.13997082e-04, -7.51313530e-07, 2.34640477e-05, 6.06731757e-07, 3.28340222e-07
        dL = a00+a01*B+a02*B**2+a03*L+a04*B*L+a05*L**2
        dB = a10+a11*B+a12*B**2+a13*L+a14*B*L+a15*L**2
    return lon_tky + dL, lat_tky + dB


def xy2ll(inputPath: Path, outputPath: Path):
    json_open = open(inputPath, 'r')
    json_load = json.load(json_open)

    with open(outputPath, 'w') as f:
        ls = []
        features = []

        for elem in json_load['features']:
            newCoordinates = []

            for lnglat in elem['geometry']['coordinates']:
                lng, lat = tky2wgs_approx(lnglat[0], lnglat[1], )
                newCoordinates.append([lng, lat])

            featureStr = {
                "type": "Feature",
                "geometry": {
                    "type": "Polygon",
                    "coordinates": newCoordinates
                },
                "properties": elem['properties']
            }
            features.insert(len(ls), featureStr)

        featureCollection = {
            "type": "FeatureCollection",
            "features": features
        }
        featureCollection = json.dumps(featureCollection, indent=4, ensure_ascii=False)
        ls.insert(len(ls), featureCollection)

        f.writelines(ls)

def main():
    outputPath = Path(__file__).parent/'../../data/out-farmland-converted.geojson'
    inputPath = Path(__file__).parent/'../../data/out-farmland-edited.geojson'
    xy2ll(inputPath, outputPath)

if __name__ == '__main__':
    main()

File: javaScript/shpToGeojson/convertGeodetic.js
const fs = require('fs');
const cblxy = require('./cblxy.js');

// japan to world
function get_world_latlng(lat, lng) {
  var world_lat = lat - lat * 0.00010695 + lng * 0.000017464 + 0.0046017,
  world_lng = lng - lat * 0.000046038 - lng * 0.000083043 + 0.010040;
  return {
    lat: world_lat,
    lng: world_lng
  };
}


// world to japan
function get_japan_latlng(lat, lng) {
  var japan_lat = lat * 1.000106961 - lng * 0.000017467 - 0.004602017,
  japan_lng = lng * 1.000083049 + lat * 0.000046047 - 0.010041046;
  return {
    lat: japan_lat,
    lng: japan_lng
  };
}


// initialize the geojson
console.log('{\n\t"type": "FeatureCollection",\n\t "features": [');

// convert geojson
const farmlandObject = JSON.parse(fs.readFileSync('./out-farmland.geojson', 'utf8'));

var count = 0;

farmlandObject.features.forEach((obj) => {
  if (count != 0) console.log('\t\t,');
  count++;
  console.log('\t\t{\n\t\t\t"type": "Feature",\n\t\t\t"properties": {\n\t\t\t\t"stroke": "#000",\n\t\t\t\t"stroke-opacity": 0.8,\n\t\t\t\t"stroke-width": 2,\n\t\t\t\t"fill": "#ff0",\n\t\t\t\t"fill-opacity": 0.5\n\t\t\t},\n\t\t\t"geometry": {\n\t\t\t\t"type": "Polygon",\n\t\t\t\t"coordinates": [');
  console.log('\t\t\t\t\t[');
  var list = obj.geometry.coordinates[0];
  for (var i in list) {
    var y = list[i][0];
    var x = list[i][1];
    var latlng = cblxy.xy2bl(x, y);
    console.log(latlng);
    var newLatlng = get_world_latlng(latlng[0], latlng[1]);
    console.log('\t\t\t\t\t\t[\n\t\t\t\t\t\t\t' + newLatlng.lng + ',\n\t\t\t\t\t\t\t' + newLatlng.lat + '\n\t\t\t\t\t\t]');
  }
  console.log('\t\t\t\t\t]');
  console.log('\t\t\t\t]\n\t\t\t}\n\t\t}');
  // obj.coordinates.forEach((elem) => {
  //   console.log('\t\t\t\t[');
  //   elem.forEach((elemelem) => {
  //     var lng = elemelem[0];
  //     var lat = elemelem[1];
  //     var latlng = get_world_latlng(lat, lng);
  //     console.log('\t\t\t\t\t[\n\t\t\t\t\t\t' + latlng.lng + ',\n\t\t\t\t\t\t' + latlng.lat + '\n\t\t\t\t\t]');
  //   });
  //   console.log('\t\t\t\t]\n');
  // });
  // console.log('\t\t\t\t]\n\t\t\t}\n\t\t}');
});

console.log('\t]\n}');
